# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediasink/DecodedStream.cpp
# Commit: 59c3902019e2
# Full Hash: 59c3902019e217e2fc118df14a50f50328030e13
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-11-19 21:20:10
# Regressor Bug: 1172394
# File Overlap Count: 1
# Description:
#   Bug 1585570 - Only create promises if we decide to create DecodedStreamData. r=padenot
#   
#   Prior to this patch, promises were always created. Since bug 1172394, creating
#   DecodedStreamData is conditional on whether the appropriate output tracks are
#   available. If promises are created and output tracks are *not* available, the
# ==============================================================================

diff -r 72fd29e8c8ee -r 59c3902019e2 dom/media/mediasink/DecodedStream.cpp
--- a/dom/media/mediasink/DecodedStream.cpp	Tue Nov 19 14:08:40 2019 +0000
+++ b/dom/media/mediasink/DecodedStream.cpp	Tue Nov 19 13:56:38 2019 +0000
@@ -285,6 +285,8 @@
   const RefPtr<ProcessedMediaTrack> mVideoOutputTrack;
   const RefPtr<MediaInputPort> mAudioPort;
   const RefPtr<MediaInputPort> mVideoPort;
+  const RefPtr<DecodedStream::EndedPromise> mAudioEndedPromise;
+  const RefPtr<DecodedStream::EndedPromise> mVideoEndedPromise;
   const RefPtr<DecodedStreamGraphListener> mListener;
 };
 
@@ -314,6 +316,8 @@
       mVideoPort((mVideoOutputTrack && mVideoTrack)
                      ? mVideoOutputTrack->AllocateInputPort(mVideoTrack)
                      : nullptr),
+      mAudioEndedPromise(aAudioEndedPromise.Ensure(__func__)),
+      mVideoEndedPromise(aVideoEndedPromise.Ensure(__func__)),
       // DecodedStreamGraphListener will resolve these promises.
       mListener(MakeRefPtr<DecodedStreamGraphListener>(
           mAudioTrack, std::move(aAudioEndedPromise), mVideoTrack,
@@ -411,12 +415,11 @@
   ConnectListener();
 
   class R : public Runnable {
-    typedef MozPromiseHolder<MediaSink::EndedPromise> Promise;
-
    public:
     R(PlaybackInfoInit&& aInit,
       nsTArray<RefPtr<ProcessedMediaTrack>> aOutputTracks,
-      Promise&& aAudioEndedPromise, Promise&& aVideoEndedPromise)
+      MozPromiseHolder<MediaSink::EndedPromise>&& aAudioEndedPromise,
+      MozPromiseHolder<MediaSink::EndedPromise>&& aVideoEndedPromise)
         : Runnable("CreateDecodedStreamData"),
           mInit(std::move(aInit)),
           mOutputTracks(std::move(aOutputTracks)),
@@ -460,15 +463,13 @@
    private:
     PlaybackInfoInit mInit;
     const nsTArray<RefPtr<ProcessedMediaTrack>> mOutputTracks;
-    Promise mAudioEndedPromise;
-    Promise mVideoEndedPromise;
+    MozPromiseHolder<MediaSink::EndedPromise> mAudioEndedPromise;
+    MozPromiseHolder<MediaSink::EndedPromise> mVideoEndedPromise;
     UniquePtr<DecodedStreamData> mData;
   };
 
   MozPromiseHolder<DecodedStream::EndedPromise> audioEndedHolder;
-  mAudioEndedPromise = audioEndedHolder.Ensure(__func__);
   MozPromiseHolder<DecodedStream::EndedPromise> videoEndedHolder;
-  mVideoEndedPromise = videoEndedHolder.Ensure(__func__);
   PlaybackInfoInit init{aStartTime, aInfo};
   nsCOMPtr<nsIRunnable> r = new R(
       std::move(init), nsTArray<RefPtr<ProcessedMediaTrack>>(mOutputTracks),
@@ -478,6 +479,8 @@
   mData = static_cast<R*>(r.get())->ReleaseData();
 
   if (mData) {
+    mAudioEndedPromise = mData->mAudioEndedPromise;
+    mVideoEndedPromise = mData->mVideoEndedPromise;
     mOutputListener = mData->OnOutput().Connect(mOwnerThread, this,
                                                 &DecodedStream::NotifyOutput);
     SendData();
